### PR TITLE
feat(system): Add new_issue count health check for release threshold

### PR DESCRIPTION
**Pull request by ykamo001**
Add logic to determine is_healthy for release thresholds. Uses `Group` Django model to aggregate the data, and uses bulk query to avoid N+1 queries.

Resolves: https://github.com/getsentry/team-enterprise/issues/18

### CODE REVIEW

File: src/sentry/api/endpoints/release_thresholds/health_checks/is_new_issues_count_healthy.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    # Each resultant threshold should be a dictionary with the 2 columns
+    results = q.values(NEW_ISSUES_UNIQUE_KEY_COLUMN, NEW_ISSUES_COUNT_COLUMN)
+
+    # Update the enriched_thresholds with the results
+    for result in results:
+        unique_key = result[
+            NEW_ISSUES_UNIQUE_KEY_COLUMN
+        ]  # Get the unique key from the result dictionary
+        value = result[NEW_ISSUES_COUNT_COLUMN]  # Get the count value
+        ethreshold = queries[unique_key]
```
Comment (nhsiehgit) [inline src/sentry/api/endpoints/release_thresholds/health_checks/is_new_issues_count_healthy.py:53]: If we need some sort of key to key results off of, we should be able to utilize the threshold id 🤷

File: src/sentry/api/endpoints/release_thresholds/health_checks/is_new_issues_count_healthy.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        if ethreshold["threshold_type"] != ReleaseThresholdType.NEW_ISSUE_COUNT_STR:
+            continue
+
+        unique_id = str(uuid4())
+        queries[unique_id] = ethreshold
+        annotation = get_is_new_issues_count_healthy_query(ethreshold, unique_id)
+        if q is None:
+            q = annotation
+        else:
+            q = q.union(annotation, all=True)
```
[the remark this answers is not in the document]
Comment (nhsiehgit) [reply src/sentry/api/endpoints/release_thresholds/health_checks/is_new_issues_count_healthy.py:38]: My SQLfu is pretty terrible though 😬 
I would def consult someone more familiar with django's query language 😅

File: src/sentry/api/endpoints/release_thresholds/health_checks/is_new_issues_count_healthy.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    It will group by the unique key.
+    The expectation is for the caller to only have to use the `unique_key` and `count` fields.
+    The reason for this function is to help bulk query the results to the datastore and prevent N+1 queries.
+    """
+    release_id = release_threshold["release_id"]
+    project_id = release_threshold["project_id"]
+    time_range = (release_threshold["start"], release_threshold["end"])
+
+    query = Group.objects.filter(
+        project__id=project_id, first_release__id=release_id, first_seen__range=time_range
```
Comment (nhsiehgit) [inline src/sentry/api/endpoints/release_thresholds/health_checks/is_new_issues_count_healthy.py:73]: gonna carry over the conversation from the [previous thread](https://github.com/getsentry/sentry/pull/62913#discussion_r1450692390)

**Review by nhsiehgit — COMMENTED** [review]
comments

File: src/sentry/api/endpoints/release_thresholds/health_checks/is_new_issues_count_healthy.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    Updates the release thresholds is_healthy for the given iterable.
+    This function leverages bulk querying to update the values of the release thresholds.
+    Each release threshold is considered to be a unique threshold to determine its own health.
+    """
+    queries = {}
+    q: Optional[QuerySet] = None
+    # Loop through each enriched_threshold and get the query that we need to aggregate
+    for ethreshold in release_thresholds:
+        # Ignore any incorrect thresholds that might have slipped in
+        if ethreshold["threshold_type"] != ReleaseThresholdType.NEW_ISSUE_COUNT_STR:
```
[the remark this answers is not in the document]
Comment (ykamo001) [reply src/sentry/api/endpoints/release_thresholds/health_checks/is_new_issues_count_healthy.py:29]: What is the hesitation for adding extra protection?

[the remark this answers is not in the document]
Comment (ykamo001) [reply src/sentry/api/endpoints/release_thresholds/health_checks/is_new_issues_count_healthy.py:29]: Referencing pure functions: https://github.com/getsentry/sentry/pull/63081#discussion_r1450716165, wouldn't having this protection be in line with that? If we assume/expect to have this filtered already, it becomes less "pure" and stateful.

File: src/sentry/api/endpoints/release_thresholds/health_checks/is_new_issues_count_healthy.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ykamo001) [reply src/sentry/api/endpoints/release_thresholds/health_checks/is_new_issues_count_healthy.py:38]: What is the main hesitation with the way it's currently queried? Is it that we're using some sort of unique key?

File: src/sentry/api/endpoints/release_thresholds/health_checks/is_new_issues_count_healthy.py (hunk 1 above)
Comment (ykamo001) [reply src/sentry/api/endpoints/release_thresholds/health_checks/is_new_issues_count_healthy.py:53]: Yeah we can def re-use release_threshold id

**Comment by ykamo001** [comment]
closing this in favor of: https://github.com/getsentry/sentry/pull/63487

**Pull request closed without merging**